OTLP: add support for counters and gauges

felixbarny opened the pull request:
Part of https://github.com/elastic/elasticsearch/pull/133057

felixbarny committed:
OTLP: add support for counters and gauges

elasticsearchmachine commented:
Pinging @elastic/es-storage-engine (Team:StorageEngine)

Commit:
[CI] Auto commit changes from spotless

martijnvg reviewed (approved):
I'm not familiar with Otel related domain classes,  but the integration into rest and transport action LGTM.

kkrik-es commented on x-pack/plugin/otel-data/src/main/java/org/elasticsearch/xpack/oteldata/otlp/OTLPMetricsTransportAction.java:
```diff
@@ -62,7 +80,112 @@ public OTLPMetricsTransportAction(
+                        // If the processing of the request fails,
+                        // the server MUST respond with appropriate HTTP 4xx or HTTP 5xx status code.
+                        new MetricsResponse(
+                            RestStatus.INTERNAL_SERVER_ERROR,
```
@martijnvg I don't think we want to return 500 here?

kkrik-es commented on x-pack/plugin/otel-data/src/main/java/org/elasticsearch/xpack/oteldata/otlp/datapoint/DataPoint.java:
```diff
@@ -0,0 +1,156 @@
+     *
+     * @return the timestamp in nanoseconds
+     */
+    long getTimestampUnixNano();
```
Do we need nanos for metrics? Or is this part of otlp?

kkrik-es commented on x-pack/plugin/otel-data/src/main/java/org/elasticsearch/xpack/oteldata/otlp/datapoint/DataPointGroupingContext.java:
```diff
@@ -0,0 +1,316 @@
+    private ResourceGroup getOrCreateResourceGroup(ResourceMetrics resourceMetrics) {
+        TsidBuilder resourceTsidBuilder = ResourceTsidFunnel.forResource(byteStringAccessor, resourceMetrics);
+        Hash128 resourceHash = resourceTsidBuilder.hash();
+        ResourceGroup resourceGroup = resourceGroups.get(resourceHash);
```
Nit: use `computeIfAbsent`?

kkrik-es commented on x-pack/plugin/otel-data/src/main/java/org/elasticsearch/xpack/oteldata/otlp/datapoint/DataPointGroupingContext.java:
```diff
@@ -0,0 +1,316 @@
+        private final ResourceGroup resourceGroup;
+        private final InstrumentationScope scope;
+        private final ByteString scopeSchemaUrl;
+        // index -> timiestamp -> dataPointGroupHash -> DataPointGroup
```
timestamp uses a Hash128?

kkrik-es commented:
Let's split the data point and tsid logic to separate PRs and add unit tests.

felixbarny commented on x-pack/plugin/otel-data/src/main/java/org/elasticsearch/xpack/oteldata/otlp/datapoint/DataPoint.java:
```diff
@@ -0,0 +1,156 @@
+        }
+
+        @Override
+        public boolean isValid(Set<String> messages) {
```
```suggestion
        public boolean isValid(Set<String> errors) {
```

felixbarny committed:
Apply suggestions from code review

Co-authored-by: Kostas Krikellas <131142368+kkrik-es@users.noreply.github.com>

felixbarny committed:
Revert changes to action and doc builder

felixbarny committed:
Add unit tests

felixbarny committed:
Merge branch 'main' into otlp-counter-gauge

felixbarny committed:
Add unit tests for BufferedByteStringAccessor

kkrik-es commented on x-pack/plugin/otel-data/src/test/java/org/elasticsearch/xpack/oteldata/otlp/proto/BufferedByteStringAccessorTests.java:
```diff
@@ -0,0 +1,49 @@
+    private final BufferedByteStringAccessor accessor = new BufferedByteStringAccessor();
+
+    public void testAddStringDimension() {
+        String value = randomUnicodeOfLengthBetween(10, 1000);
```
Check range [1, 1000]? What about empty string?

felixbarny committed:
Add unit tests for AttributeListTsidFunnel

felixbarny committed:
Address review comments

Commit:
[CI] Auto commit changes from spotless

felixbarny committed:
Inline and remove unnecessary DataPointDimensionsTsidFunnel

felixbarny committed:
Remove scope schema_url and version from grouping

This is in line with that the ES exporter is doing and these fields aren't mapped as a dimension anyway

felixbarny committed:
Merge remote-tracking branch 'origin/main' into otlp-counter-gauge

felixbarny committed:
Merge remote-tracking branch 'felixbarny/otlp-counter-gauge' into otlp-counter-gauge

felixbarny committed:
Merge branch 'main' into otlp-counter-gauge
